### ISSUE TITLE
Variable does not get gradient

### ISSUE THREAD

**Issue by greatwall1995** [pytorch/pytorch#1760]
Is there any difference between `x` and `Variable(torch.Tensor(x.data.numpy()), requires_grad=x.requires_grad, volatile=x.volatile)` ?

In my code, I set `self.y=x` in `forward()` function. After calling `backward()` function (without parameter in it), `y.grad` is `None`. However, if I set  `self.y=Variable(torch.Tensor(x.data.numpy()), requires_grad=x.requires_grad, volatile=x.volatile)`, `y.grad` can be calculated.

It quite strange to me that the results of these two codes are different.

**UPD:**  I finally condense the code like this:
~~~~
import torch
import torch.nn as nn
from torch.autograd import Variable

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc0 = nn.Linear(3, 5)
    def forward(self, x):
        x = torch.cat(x)
        self.y = x #Variable(x.data, requires_grad=x.requires_grad, volatile=x.volatile)
        x = self.fc0(self.y)
        return x

model = Net()
x = [Variable(torch.Tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]]), requires_grad=True) for i in xrange(3)]
output = model(x)
output[0][0].backward()
print model.y.grad
~~~~

**UPD2:** And I also tried another code where the input of `forward()` function is a Variable. It still has the same problem.
~~~~
import torch
import torch.nn as nn
from torch.autograd import Variable

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc = nn.Linear(3, 5)
        self.internal = [Variable(torch.Tensor([[1, 2, 3], [1, 2, 3], [1, 2, 3]])) for i in xrange(3)]
    def forward(self, x):
        x = torch.cat([torch.mm(x, self.internal[i]) for i in xrange(3)])
        self.y = x #Variable(x.data, requires_grad=x.requires_grad, volatile=x.volatile)
        x = self.fc(self.y)
        return x

model = Net()
x = Variable(torch.Tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]]), requires_grad=True)
output = model(x)
output[0][0].backward()
print model.y.grad
~~~~

**Comment by chenzhekl** [comment]
It is better to post this question to the forum.

`self.y = x` simply makes `y` points to `x`. While `Variable(torch.Tensor(x.data.numpy()), requires_grad=x.requires_grad, volatile=x.volatile)` recreats a new `Variable` using data in `x`. It loses all computational history of `x`.

In your case, `y.grad` is not calculated because it is not a leaf node when `y = x`. Recreating a new `Variable` makes it a leaf node.

**Comment by soumith** [comment]
https://discuss.pytorch.org (for future reference)
